File: src/my_pa/contracts/v1/disclosure.py
"""The mandatory disclosure envelope.

Every successful or partial result states its scope, coverage, freshness, trust,
truncation, limitations, source references, and unavailable evidence. Counts are
scoped to the authorized enrollment and never imply global coverage
(`docs/specs`, section 8.3).
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import Field, model_validator

from my_pa.contracts.v1.base import StrictModel, UtcDatetime
from my_pa.domain.common.classification import Classification, is_cloud_eligible
from my_pa.domain.common.identifiers import IdKind, validate_identifier
from my_pa.domain.common.provenance import TrustLevel

__all__ = [
    "Coverage",
    "CoverageState",
    "Disclosure",
    "Freshness",
    "FreshnessState",
    "Scope",
    "SourceReference",
    "Truncation",
    "Trust",
]


class CoverageState(StrEnum):
    """Distinct coverage states; none of them collapses into "empty"."""

    NOT_ENROLLED = "not_enrolled"
    ELIGIBLE = "eligible"
    QUEUED = "queued"
    PROCESSED = "processed"
    PARTIALLY_PROCESSED = "partially_processed"
    UNSUPPORTED = "unsupported"
    QUARANTINED = "quarantined"
    UNAVAILABLE = "unavailable"
    STALE = "stale"
    SUPERSEDED = "superseded"


class FreshnessState(StrEnum):
    """How current the observed state is."""

    CURRENT_FOR_OBSERVED_VERSION = "current_for_observed_version"
    STALE = "stale"
    UNKNOWN = "unknown"


def _validate_ids(values: tuple[str, ...], kind: IdKind) -> tuple[str, ...]:
    for value in values:
        validate_identifier(value, kind)
    return values


class Scope(StrictModel):
    """The authorized scope a result was produced within."""

    source_ids: tuple[str, ...] = ()
    enrollment_ids: tuple[str, ...] = ()

    @model_validator(mode="after")
    def _check_identifiers(self) -> Scope:
        _validate_ids(self.source_ids, IdKind.SOURCE)
        _validate_ids(self.enrollment_ids, IdKind.ENROLLMENT)
        return self


class Coverage(StrictModel):
    """Bounded counts for the stated scope."""

    state: CoverageState
    eligible: int = Field(default=0, ge=0)
    processed: int = Field(default=0, ge=0)
    quarantined: int = Field(default=0, ge=0)
    unsupported: int = Field(default=0, ge=0)

    @model_validator(mode="after")
    def _check_totals(self) -> Coverage:
        accounted = self.processed + self.quarantined + self.unsupported
        if accounted > self.eligible:
            raise ValueError("processed, quarantined and unsupported cannot exceed eligible")
        return self


class Freshness(StrictModel):
    """When the reported state was observed."""

    observed_at: UtcDatetime
    state: FreshnessState


class Trust(StrictModel):
    """How much authority the result carries and why."""

    level: TrustLevel
    basis: tuple[str, ...] = ()


class Truncation(StrictModel):
    """Explicit truncation, never an unmarked complete-looking response."""

    is_truncated: bool = False
    reason: str | None = None
    next_cursor: str | None = None

    @model_validator(mode="after")
    def _check_consistency(self) -> Truncation:
        if self.is_truncated and not self.reason:
            raise ValueError("a truncated result must state a reason")
        if not self.is_truncated and (self.reason or self.next_cursor):
            raise ValueError("an untruncated result cannot carry truncation details")
        return self


class SourceReference(StrictModel):
    """Opaque binding to the source state a result came from."""

    source_id: str
    source_object_id: str
    version_id: str

    @model_validator(mode="after")
    def _check_identifiers(self) -> SourceReference:
        validate_identifier(self.source_id, IdKind.SOURCE)
        validate_identifier(self.source_object_id, IdKind.SOURCE_OBJECT)
        validate_identifier(self.version_id, IdKind.VERSION)
        return self


class Disclosure(StrictModel):
    """The mandatory envelope accompanying every success or partial result."""

    scope: Scope
    coverage: Coverage
    freshness: Freshness
    trust: Trust
    truncation: Truncation = Truncation()
    limitations: tuple[str, ...] = ()
    source_references: tuple[SourceReference, ...] = ()
    unavailable_evidence: tuple[str, ...] = ()
    partial_result: bool = False
    classification: Classification = Classification.PRIVATE_LOCAL
    cloud_eligible: bool = False

    @model_validator(mode="after")
    def _check_partial_is_truthful(self) -> Disclosure:
        partial_states = {
            CoverageState.PARTIALLY_PROCESSED,
            CoverageState.QUARANTINED,
            CoverageState.UNSUPPORTED,
            CoverageState.UNAVAILABLE,
            CoverageState.STALE,
        }
        if self.coverage.state in partial_states and not self.partial_result:
            raise ValueError(f"coverage state {self.coverage.state} requires partial_result=True")
        if self.truncation.is_truncated and not self.partial_result:
            raise ValueError("a truncated result must set partial_result=True")
        return self

    @model_validator(mode="after")
    def _check_cloud_eligibility(self) -> Disclosure:
        """Reject a disclosure claiming eligibility the classification denies.

        Without this the mandatory envelope could assert that restricted local
        content may leave the trust boundary, contradicting the domain rule it
        is supposed to be reporting.
        """
        if self.cloud_eligible and not is_cloud_eligible(self.classification):
            raise ValueError(f"classification {self.classification} is never cloud eligible")
        return self

File: web/src/lib/api/decode/disclosure.ts
/**
 * Runtime decode of the Python `Disclosure` known fields.
 *
 * Required known fields are strict. Unknown extra keys (`scope`,
 * `source_references`, `classification`, `cloud_eligible`, coverage counts,
 * `truncation.reason`, …) are ignored. Invalid enums and missing required
 * fields fail closed. Nothing here defaults coverage to complete or trust to
 * `"derived"`.
 */
import {
  closed,
  ignoreUnknownKeys,
  isBoolean,
  isRecord,
  isString,
  isStringArray,
  ok,
  optional,
  type DecodeResult,
} from "./primitives";

export const COVERAGE_STATES = [
  "not_enrolled",
  "eligible",
  "queued",
  "processed",
  "partially_processed",
  "unsupported",
  "quarantined",
  "unavailable",
  "stale",
  "superseded",
] as const;

export type CoverageState = (typeof COVERAGE_STATES)[number];

export const FRESHNESS_STATES = ["current_for_observed_version", "stale", "unknown"] as const;

export type FreshnessState = (typeof FRESHNESS_STATES)[number];

export const TRUST_LEVELS = ["source_original", "source_bound_derived", "model_proposed"] as const;

export type TrustLevel = (typeof TRUST_LEVELS)[number];

export interface DecodedDisclosure {
  readonly coverage: { readonly state: CoverageState };
  readonly freshness: { readonly observed_at: string; readonly state: FreshnessState };
  readonly trust: { readonly level: TrustLevel; readonly basis: readonly string[] };
  readonly truncation: { readonly is_truncated: boolean; readonly next_cursor?: string };
  readonly limitations: readonly string[];
  readonly partial_result: boolean;
}

const DISCLOSURE_KEYS = [
  "coverage",
  "freshness",
  "trust",
  "truncation",
  "limitations",
  "partial_result",
] as const;

const INVALID = "upstream_contract_invalid";

function oneOf<T extends string>(
  value: unknown,
  allowed: readonly T[],
): DecodeResult<T> {
  if (isString(value)) {
    for (const candidate of allowed) {
      if (value === candidate) return ok(candidate);
    }
  }
  return closed(INVALID, "a required field was not an allowed value");
}

function decodeCoverage(value: unknown): DecodeResult<{ state: CoverageState }> {
  if (!isRecord(value)) {
    return closed(INVALID, "disclosure coverage was missing or unreadable");
  }
  const known = ignoreUnknownKeys(value, ["state"]);
  const state = oneOf(known.state, COVERAGE_STATES);
  if (!state.ok) return state;
  return ok({ state: state.value });
}

function decodeFreshness(
  value: unknown,
): DecodeResult<{ observed_at: string; state: FreshnessState }> {
  if (!isRecord(value)) {
    return closed(INVALID, "disclosure freshness was missing or unreadable");
  }
  const known = ignoreUnknownKeys(value, ["observed_at", "state"]);
  if (!isString(known.observed_at)) {
    return closed(INVALID, "a required disclosure field was missing");
  }
  const state = oneOf(known.state, FRESHNESS_STATES);
  if (!state.ok) return state;
  return ok({ observed_at: known.observed_at, state: state.value });
}

function decodeTrust(
  value: unknown,
): DecodeResult<{ level: TrustLevel; basis: readonly string[] }> {
  if (!isRecord(value)) {
    return closed(INVALID, "disclosure trust was missing or unreadable");
  }
  const known = ignoreUnknownKeys(value, ["level", "basis"]);
  const level = oneOf(known.level, TRUST_LEVELS);
  if (!level.ok) return level;
  if (!isStringArray(known.basis)) {
    return closed(INVALID, "a required disclosure field was missing");
  }
  return ok({ level: level.value, basis: known.basis });
}

function decodeTruncation(
  value: unknown,
): DecodeResult<{ is_truncated: boolean; next_cursor?: string }> {
  if (!isRecord(value)) {
    return closed(INVALID, "disclosure truncation was missing or unreadable");
  }
  const known = ignoreUnknownKeys(value, ["is_truncated", "next_cursor"]);
  if (!isBoolean(known.is_truncated)) {
    return closed(INVALID, "a required disclosure field was missing");
  }
  const cursor = optional(known.next_cursor, (present) =>
    isString(present)
      ? ok(present)
      : closed(INVALID, "a disclosure field was not the expected type"),
  );
  if (!cursor.ok) return cursor;
  return ok(
    cursor.value === undefined
      ? { is_truncated: known.is_truncated }
      : { is_truncated: known.is_truncated, next_cursor: cursor.value },
  );
}

export function decodeDisclosure(input: unknown): DecodeResult<DecodedDisclosure> {
  if (!isRecord(input)) {
    return closed(INVALID, "disclosure was missing or unreadable");
  }
  const known = ignoreUnknownKeys(input, DISCLOSURE_KEYS);
  const coverage = decodeCoverage(known.coverage);
  if (!coverage.ok) return coverage;
  const freshness = decodeFreshness(known.freshness);
  if (!freshness.ok) return freshness;
  const trust = decodeTrust(known.trust);
  if (!trust.ok) return trust;
  const truncation = decodeTruncation(known.truncation);
  if (!truncation.ok) return truncation;
  if (!isStringArray(known.limitations)) {
    return closed(INVALID, "a required disclosure field was missing");
  }
  if (!isBoolean(known.partial_result)) {
    return closed(INVALID, "a required disclosure field was missing");
  }
  return ok({
    coverage: coverage.value,
    freshness: freshness.value,
    trust: trust.value,
    truncation: truncation.value,
    limitations: known.limitations,
    partial_result: known.partial_result,
  });
}
